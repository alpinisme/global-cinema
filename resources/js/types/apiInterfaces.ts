/**
 * This file contains interfaces for the objects returned from our backend apis
 * The interfaces are non-exhaustive. That is, additional data may also be transmitted
 * But these interfaces represent a minimum as well as a necessary structure
 *
 * Backend api tests should be maintained so ensure that data is delivered in this format
 */

/**
 * get requests to films/ api return an array of objects with interface Film
 */
export interface Film {
    id: string;
    title: string;
    matches: Film[];
    isNew: boolean;
}

/**
 * get requests to theaters/ api return an array of objects with interface Film
 */
export interface Theater {
    name: string;
    id: string;
}

/**
 * get requests to screenings/ api return an array of objects with interface Film
 */
export interface Screening {
    id: number;
    theater_id: string;
    film_id: string;
}

/**
 * get requests to instructor/ api return an array of objects with interface Student
 */
export interface Student {
    info: {
        email: string;
        name: string;
        id: string;
    };
    datesCompleted: string[];
}
