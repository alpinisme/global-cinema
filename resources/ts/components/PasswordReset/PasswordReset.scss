@use './../../../sass/variables';

$main: variables.$main;

.reset {
    color: $main;
    border: none;
    background: none;
    margin-top: 1rem;
}
