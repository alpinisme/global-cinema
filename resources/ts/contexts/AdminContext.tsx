import { createContext } from 'react';
import type { User, City } from '../types/api';

export interface AdminContextData {
    users: User[];
    cities: City[];
}

const AdminContext = createContext({} as AdminContextData);

export default AdminContext;
