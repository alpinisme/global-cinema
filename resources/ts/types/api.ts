/**
 * This file contains interfaces for the objects returned from our backend apis
 * The interfaces are non-exhaustive. That is, additional data may also be transmitted
 * But these interfaces represent a minimum as well as a necessary structure
 *
 * Backend api tests should be maintained so ensure that data is delivered in this format
 */

/**
 * get requests to `films/` api return an array of objects with interface `Film`
 */
export interface Film {
    id: number;
    title: string;
    year: number;
}

/**
 * get requests to `theaters/` api return an array of objects with interface `Film`
 */
export interface Theater {
    name: string;
    id: number;
    city_id: number;
}

/**
 * get requests to `screenings/` api return an array of objects with interface `Screening`
 */
export interface Screening {
    id?: number;
    theater_id: number;
    film_id: number;
    city_id: number;
    date: string;
}

/**
 * get requests to `instructor/` api return an array of objects with interface `Student`
 */
export interface Student {
    info: {
        email: string;
        name: string;
        id: number;
    };
    datesCompleted: string[];
}

/**
 * get requests to `users/` api return an array of objects with interface `User`
 */
export interface User {
    email: string;
    name: string;
    id: number;
    role: string;
}
