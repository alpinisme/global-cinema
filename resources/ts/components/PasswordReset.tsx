import React, { ReactElement, useContext } from 'react';
import { User } from '../types/api';
import Autosuggest from './Autosuggest';
import Clipboard from './Clipboard';
import { usePostRequest } from '../utils/hooks';
import AdminContext from '../contexts/AdminContext';

const PasswordReset = (): ReactElement => {
    const [passResetResult, makePassResetRequest] = usePostRequest<string, string>();
    const { users } = useContext(AdminContext);

    const suggestUsersConfig = {
        label: 'User',
        keys: ['name', 'email'],
        displayMatch: (u: User) => `${u.name} (${u.email})`,
        options: users ?? [],
    };

    if (passResetResult.data) {
        return (
            <>
                Reset link:
                <Clipboard content={passResetResult.data} />
            </>
        );
    }

    return (
        <Autosuggest
            config={suggestUsersConfig}
            handleSubmit={id => makePassResetRequest(`/password/reset/${id}`, '')}
        />
    );
};

export default PasswordReset;
