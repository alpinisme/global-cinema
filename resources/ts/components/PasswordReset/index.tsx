import React, { ReactElement, useContext } from 'react';
import { User } from '../../types/api';
import Autosuggest from '../Autosuggest';
import Clipboard from '../Clipboard';
import { usePostRequest } from '../../hooks/requestHooks';
import { AdminContext } from '../../contexts/AdminContext';
import styles from './PasswordReset.scss';

// TODO: allow more than one password reset (add action to button)

const PasswordReset = (): ReactElement => {
    const [passResetResult, makePassResetRequest] = usePostRequest<string, string>();
    const { users } = useContext(AdminContext);

    const suggestUsersConfig = {
        label: 'User',
        keys: ['name', 'email'],
        displayMatch: (u: User) => `${u.name} (${u.email})`,
        options: users ?? [],
    };

    if (passResetResult.data) {
        return (
            <>
                Reset link:
                <Clipboard content={passResetResult.data} />
                <button className={styles.reset}>reset another password?</button>
            </>
        );
    }

    return (
        <Autosuggest
            config={suggestUsersConfig}
            handleSubmit={id => makePassResetRequest(`/password/reset/${id}`, '')}
        />
    );
};

export default PasswordReset;
