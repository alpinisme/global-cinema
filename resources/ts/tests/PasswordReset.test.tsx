import React from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import PasswordReset from '../components/PasswordReset';
import AdminContext from '../contexts/AdminContext';
import { City } from '../types/api';

let el: ReactTestRenderer;

const context = {
    cities: [] as City[],
    users: [{ name: 'Matthew Nelson', email: 'email@email.com', id: 1, role: 'admin' }],
};

beforeEach(() => {
    el = create(
        <AdminContext.Provider value={context}>
            <PasswordReset />
        </AdminContext.Provider>
    );
});

describe('Password Reset', () => {
    it('renders correctly', () => {
        expect(el).toMatchSnapshot();
    });

    it('contains no buttons before user provides input', () => {
        const inputs = el.root.findAllByType('button');
        expect(inputs.length).toBe(0);
    });

    it('contains a button with correct name if user begins typing a search term', () => {
        const input = el.root.findByType('input');
        expect(input.children.includes('Matthew Nelson (email@email.com)')).toBe(false);
        act(() => input.props.onChange({ target: { value: 'Matt' } }));

        const button = el.root.findByType('button');
        expect(button.children.includes('Matthew Nelson (email@email.com)')).toBe(true);
    });

    it('contains no buttons if user deletes their search term', () => {
        const input = el.root.findByType('input');
        act(() => input.props.onChange({ target: { value: 'Matt' } }));
        act(() => input.props.onChange({ target: { value: '' } }));

        const buttons = el.root.findAllByType('button');
        expect(buttons.length).toBe(0);
    });
});
