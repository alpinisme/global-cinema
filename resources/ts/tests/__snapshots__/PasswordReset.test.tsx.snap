// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password Reset renders correctly 1`] = `
Array [
  <div>
    <label
      className="label"
      htmlFor="User"
    >
      User
      :
    </label>
    <input
      className="input"
      name="User"
      onChange={[Function]}
      type="text"
      value=""
    />
    <p>
      As you begin typing, possible matches will appear below.
    </p>
  </div>,
  "",
]
`;
